include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -53,6 +57,47 @@ __pattern_walk_n(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _Function
     }
 }
 
+#if _ONEDPL_CPP20_RANGES_PRESENT
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_for_each
+//---------------------------------------------------------------------------------------------------------------------
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun>
+auto
+__pattern_for_each(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj)
+{
+    auto __f_1 = 
+        [__f, __proj](auto&& __val) { std::invoke(__f, std::invoke(__proj, std::forward<decltype(__val)>(__val)));};
+    oneapi::dpl::__internal::__ranges::__pattern_walk_n(__tag, std::forward<_ExecutionPolicy>(__exec), __f_1,
+                                                            oneapi::dpl::views::all(std::forward<_R>(__r)));
+
+    return std::ranges::borrowed_iterator_t<_R>(std::ranges::begin(__r) + std::ranges::size(__r));
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+// pattern_transform
+//---------------------------------------------------------------------------------------------------------------------
+template<typename _BackendTag, typename _ExecutionPolicy, typename _InRange, typename _OutRange, typename _F,
+         typename _Proj>
+auto
+__pattern_transform(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _InRange&& __in_r, _OutRange&& __out_r,
+                    _F __op, _Proj __proj)
+{
+    auto __unary_op = [=](auto&& __val) -> decltype(auto)
```

Comment:
here we have `decltype(auto)` as a return type, while in `find_if` we don't What is the reason?

I am not judging what is correct and what is not. Just trying to understand.

I could imagine that we are doing that because `transform` function can return anything, while `find_if` predicate returns `bool`.

Reply:
we write `-> decltype(auto)` possibility to have `lvalue `after unary transformation is applied.
Without `-> decltype(auto)`:  `error: expression is not assignable`

Reply:
I added `-> decltype(auto)` because there were issues in the test.
Unfortunately, i cannot remember exact case for now.

Reply:
I've removed `decltype(auto)` and verified the changes with g++12 (Linux), icpx 2024.2 (Linux), icx 2024.2 (Windows), cl 19.34. 
I guess it is not necessary, and there was a confusion with e.g. another fix.